## Rerun of acquiring DOIs using the Crossref API with the python library 'habanero' 
(conda installed v0.6)
check with first ten 



code for basic install and a simple search of the works using a query and a filter



In [153]:
import pandas as pd
from habanero import Crossref, exceptions
import time

### First bring in the existing list of publictaions (using the columns needed)

In [154]:
df1 = pd.read_csv('./A3out_CrossRef_BRC_FuzzyMatched_07Aug19.csv', index_col=0)
df1.shape

(2378, 18)

In [155]:
df1.head(2)

DOI    ID  \
0    10.1186/s12881-014-0095-4  1125   
1  10.1183/13993003.00321-2016  1996   

                                            complete  \
0  &amp; , fenwick al, goos jac, rankin j, lord h...   
1  , pattinson kt, turner mr. a wider pathologica...   

                                    crossref_API_out  \
0  {'title': ['Apparently synonymous substitution...   
1  {'title': ['A wider pathological network under...   

                                      csv_post_title  \
0  apparently synonymous substitutions in fgfr2 a...   
1  a wider pathological network underlying breath...   

                       DOI_API  \
0    10.1186/s12881-014-0095-4   
1  10.1183/13993003.00321-2016   

                                               title  \
0  ['Apparently synonymous substitutions in FGFR2...   
1  ['A wider pathological network underlying brea...   

                                           title_API  title_match  DOI_match  \
0  Apparently synonymous substitutions in FGFR2af...        False       True   
1  A wider pathological network underlying breath...         True       True   

                                   CrossRefTitle_low  DOI_fuzzyRatio  \
0  apparently synonymous substitutions in fgfr2af...           100.0   
1  a wider pathological network underlying breath...           100.0   

   TitleMatch01  Title_fuzzyRatio  Title_partialRatio  Complete_fuzzyRatio  \
0             0              99.0                99.0                 51.0   
1             1             100.0               100.0                 73.0   

   Complete_partialRatio  good_list  
0                   99.0       True  
1                  100.0       True

In [156]:
#make dataframe from only those missing DOIs
#df1[~df1.doi.isna()]  gives those with DOIs (1440 as reported)
df1[df1.good_list==False].shape


(379, 18)

### Therefore identifying that 353 of the 2378 publications without good DOI matches
### _however, we don't know any of them are correct_ so it will best to acquire all top matches from CrossRef


In [157]:
df1mini =df1[df1.good_list==False].iloc[5:8]
df1mini

DOI    ID  \
26      10.3389/fmicb.2015.01016  1642   
33                           NaN  1639   
55  10.1097/BOR.0b013e3283620163   563   

                                             complete  \
26  adland e, klenerman p, goulder p, matthews pc....   
33  ahmed s, brennan l, eppig j, price cc, lamar m...   
55  al-mossawi mh, ridley a, kiedel s, bowness p. ...   

                                     crossref_API_out  \
26  {'title': ['Review of cytomegalovirus coinfect...   
33  {'title': ['Comparative study between depressi...   
55  {'title': ['Gamma delta T cells in non-immune ...   

                                       csv_post_title  \
26  ongoing burden of disease and mortality from h...   
33  visuoconstructional impairment in subtypes of ...   
55  the role of natural killer cells, gamma delta ...   

                             DOI_API  \
26                  10.1002/rmv.1907   
33  10.1111/j.1442-2018.2011.00666.x   
55                   10.1002/iid3.18   

                                                title  \
26  ['Review of cytomegalovirus coinfection in HIV...   
33  ['Comparative study between depression in Kore...   
55  ['Gamma delta T cells in non-immune patients d...   

                                            title_API  title_match  DOI_match  \
26  Review of cytomegalovirus coinfection in HIV-i...        False      False   
33  Comparative study between depression in Korean...        False      False   
55  Gamma delta T cells in non-immune patients dur...        False      False   

                                    CrossRefTitle_low  DOI_fuzzyRatio  \
26  review of cytomegalovirus coinfection in hiv-i...            40.0   
33  comparative study between depression in korean...             NaN   
55  gamma delta t cells in non-immune patients dur...            42.0   

    TitleMatch01  Title_fuzzyRatio  Title_partialRatio  Complete_fuzzyRatio  \
26             0              47.0                52.0                 32.0   
33             0              47.0                56.0                 38.0   
55             0              47.0                54.0                 33.0   

    Complete_partialRatio  good_list  
26                   52.0      False  
33                   49.0      False  
55                   55.0      False

In [158]:
cr = Crossref()

In [159]:
#Crossref(api_key = "168421")
# set a mailto address
Crossref(mailto = "laurence.brown@ndcn.ox.ac.uk")

< Crossref 
URL: https://api.crossref.org
KEY: None
MAILTO: laurence.brown@ndcn.ox.ac.uk
>

In [160]:
#cr.filter_details()

In [161]:
'''['DOI', 'URL', 'alternative-id', 'author','indexed', 'is-referenced-by-count',
       'language', 'license', 'link',  'published-online', 'published-print', 'publisher',
       'reference', 'reference-count', 'references-count', 'relation', 'score',
       'short-container-title', 'source', 'title', 'type', 'update-to']
'''
#select=['title', 'link','URL', 'author','is-referenced-by-count','alternative-id', 'author','indexed',],

"['DOI', 'URL', 'alternative-id', 'author','indexed', 'is-referenced-by-count',\n       'language', 'license', 'link',  'published-online', 'published-print', 'publisher',\n       'reference', 'reference-count', 'references-count', 'relation', 'score',\n       'short-container-title', 'source', 'title', 'type', 'update-to']\n"

In [162]:
crossref_test =cr.works(query=df1mini.complete[26]
             ,select=['title','DOI'],
                      filter={'from_pub_date':'2011-01-01','until_pub_date':'2018-12-31','type':'journal-article'},
                      limit=1)

In [163]:
crossref_test['message']['items'][0]['DOI'],crossref_test['message']['items'][0]['title'][0]

('10.3389/fmicb.2015.01016',
 'Ongoing burden of disease and mortality from HIV/CMV coinfection in Africa in the antiretroviral therapy era')

In [164]:
def CrossRefDoiRetry(df):
    try:
        search = cr.works(query=df.complete
             ,select=['title','DOI'],filter={'from_pub_date':'2011-01-01','until_pub_date':'2018-12-31','type':'journal-article'},
                      limit=1)
        time.sleep(1)
        return search['message']['items'][0]['DOI'], search['message']['items'][0]['title'][0]
    except:
        return None
    finally:
        time.sleep(1)
        pass

In [165]:
#df1x.columns


In [166]:
df1mini[['retry_API','retry_title']]= df1mini.apply(CrossRefDoiRetry, axis=1, result_type='expand')

In [167]:
df1mini

DOI    ID  \
26      10.3389/fmicb.2015.01016  1642   
33                           NaN  1639   
55  10.1097/BOR.0b013e3283620163   563   

                                             complete  \
26  adland e, klenerman p, goulder p, matthews pc....   
33  ahmed s, brennan l, eppig j, price cc, lamar m...   
55  al-mossawi mh, ridley a, kiedel s, bowness p. ...   

                                     crossref_API_out  \
26  {'title': ['Review of cytomegalovirus coinfect...   
33  {'title': ['Comparative study between depressi...   
55  {'title': ['Gamma delta T cells in non-immune ...   

                                       csv_post_title  \
26  ongoing burden of disease and mortality from h...   
33  visuoconstructional impairment in subtypes of ...   
55  the role of natural killer cells, gamma delta ...   

                             DOI_API  \
26                  10.1002/rmv.1907   
33  10.1111/j.1442-2018.2011.00666.x   
55                   10.1002/iid3.18   

                                                title  \
26  ['Review of cytomegalovirus coinfection in HIV...   
33  ['Comparative study between depression in Kore...   
55  ['Gamma delta T cells in non-immune patients d...   

                                            title_API  title_match  DOI_match  \
26  Review of cytomegalovirus coinfection in HIV-i...        False      False   
33  Comparative study between depression in Korean...        False      False   
55  Gamma delta T cells in non-immune patients dur...        False      False   

                                    CrossRefTitle_low  DOI_fuzzyRatio  \
26  review of cytomegalovirus coinfection in hiv-i...            40.0   
33  comparative study between depression in korean...             NaN   
55  gamma delta t cells in non-immune patients dur...            42.0   

    TitleMatch01  Title_fuzzyRatio  Title_partialRatio  Complete_fuzzyRatio  \
26             0              47.0                52.0                 32.0   
33             0              47.0                56.0                 38.0   
55             0              47.0                54.0                 33.0   

    Complete_partialRatio  good_list                     retry_API  \
26                   52.0      False      10.3389/fmicb.2015.01016   
33                   49.0      False    10.1016/j.jalz.2014.05.929   
55                   55.0      False  10.1097/bor.0b013e3283620163   

                                          retry_title  
26  Ongoing burden of disease and mortality from H...  
33  VISUOCONSTRUCTIONAL IMPAIRMENT IN SUBTYPES OF ...  
55  The role of natural killer cells, gamma delta ...

In [168]:
df_retry =df1[df1.good_list==False]

In [169]:
df_retry[['DOI_retry_API','title_retry_API']]= df_retry.apply(CrossRefDoiRetry, axis=1, result_type='expand')

C:\Users\labrown\Miniconda3\envs\biblio_tools\lib\site-packages\pandas\core\frame.py:3391: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[k1] = value[k2]


In [178]:
df_retry.to_csv('./A4out_badlist_API_retry(07Aug19).csv')

In [179]:
df_retry

DOI    ID  \
5                                       NaN  1312   
7     10.1016/j.atherosclerosis.2012.03.036    62   
10                                      NaN   168   
20         10.1016/j.fertnstert.2014.12.123  1920   
21                                      NaN  1023   
26                 10.3389/fmicb.2015.01016  1642   
33                                      NaN  1639   
55             10.1097/BOR.0b013e3283620163   563   
72                       10.3233/jpd-120126   490   
74                                      NaN    66   
77           10.1523/jneurosci.2317-15.2015  1640   
85          10.1016/j.expneurol.2015.12.004  1746   
88          10.4067/S0370-41062014000400004  1550   
97                                      NaN    33   
131                                     NaN   135   
133                       10.1111/ijs.12618  1599   
135                                     NaN   645   
138              10.1016/j.jhep.2013.09.002   842   
141                                     NaN   910   
142                    10.1093/brain/awu040   980   
144             10.1101/cshperspect.a017293  1212   
145                       10.1038/srep29784  2159   
151            10.1016/j.eururo.2015.04.036  1461   
162                                     NaN  1512   
170                                     NaN  2196   
186              10.1016/j.jacc.2016.11.068  2349   
201            10.1136/heartjnl-2015-309166  1866   
204              10.1016/j.wneu.2015.08.065  1627   
205            10.1227/NEU.0000000000000321  1016   
206              10.1016/j.jocn.2015.04.005  1486   
...                                     ...   ...   
2181                     10.1117/12.2038377   901   
2194                    10.1038/jid.2015.63  1535   
2198             10.1016/j.jinf.2013.09.013   707   
2204                                    NaN  2217   
2211              10.1016/j.cmi.2016.10.014  2307   
2219          10.1016/S2213-2600(13)70245-5   800   
2222                      10.1111/jvh.12428  1522   
2225                                    NaN   397   
2230     10.1016/B978-0-444-63432-0.00009-8  2140   
2231               10.1136/jnnp-2015-312601  2139   
2257                                    NaN   610   
2258                                    NaN   963   
2275             10.1016/j.jacc.2016.02.026  1873   
2280                      10.2217/epi.15.93  1698   
2282           10.1371/journal.ppat.1002874   142   
2285            10.1136/emermed-2011-200499   394   
2288           10.1371/journal.pgen.1005378  1669   
2289              10.1186/s12872-016-0355-7  2194   
2304                                    NaN   433   
2316          10.1016/j.vaccine.2013.04.036   514   
2325              10.1016/j.jim.2013.03.001   390   
2329              10.1007/s40279-013-0128-8  1271   
2334                                    NaN  1740   
2340           10.1016/j.ijcard.2015.08.211  1568   
2341                     10.4161/viru.21189   130   
2347                                    NaN   355   
2348                                    NaN   637   
2350               10.1136/jnnp-2014-308736  1158   
2354                      10.1002/mds.26072  1377   
2364               10.1093/eurheartj/ehv172  1909   

                                               complete  \
5     g w, rane    , harston g w, rane n, shaya g, t...   
7     kylintireas i,  craig s,  nethononda r,  kohle...   
10    pagnamenta at,  murray je,  yoon g, akha es,  ...   
20    åukaszuk k, pukszta s, wells d, cybulska c, l...   
21    abdel-hady m, bansal s, davison sm, brown m, t...   
26    adland e, klenerman p, goulder p, matthews pc....   
33    ahmed s, brennan l, eppig j, price cc, lamar m...   
55    al-mossawi mh, ridley a, kiedel s, bowness p. ...   
72    antoniades c, xuc z, carpenter rhs, barker r. ...   
74    antoniades ca, carpenter rhs. making neurology...   
77    antoniades ca, rebelo p, kennard c, aziz tz, g...   
85    anzak a, tan h, pogosyan a, khan s, jav

In [180]:
df1_out = df1.merge(df_retry, how='left',)

In [181]:
df1_out.to_csv('./A4out_CrossRef_FuzzyMatched_with_retry_07Aug19.csv')

### Note:  The 'badlist' references, with additional information from Crossref, were then manually checked, and this information was then cleaned to return a 'curated DOI' for these references.

##### File names :  A5_badlist_checkedafter_API_retry(08Aug19).xls     and    A6in_currated_DOIs.csv.